### PR TITLE
POC: tRPC for client/server communication

### PR DESCRIPTION
**Pull request by dgca**
This is a POC that demos using tRPC to handle client/server communication. If you wanna test it out, check out the branch, `yarn`, and `yarn start:dev`. The Accounts  view has the demo component. It should Just Work™

Screenshot 2023-09-28 at 11 33 05 PM

### CODE REVIEW

File: electron/api/index.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import z from 'zod'
```
Comment (dgca) [inline electron/api/index.ts:1]: This file is basically the API routes. Here there are three procedures, all in one file for simplicity.

File: electron/api/index.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+const ee = new EventEmitter()
+const t = initTRPC.create({ isServer: true })
+
+let count = 1
+function emitEveryFiveSecs() {
+  setInterval(() => {
+    ee.emit('demo-event', `Count is ${count++}`)
+  }, 5000)
+}
+emitEveryFiveSecs()
```
Comment (dgca) [inline electron/api/index.ts:16]: This just emits an event every 5 seconds, we'll use this to demo subscriptions.

File: electron/api/index.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+export const router = t.router({
+  greeting: t.procedure.input(z.object({ name: z.string() })).query(req => {
+    const { input } = req
+
+    return {
+      text: `Hello ${input.name}` as const,
+    }
+  }),
```
Comment (dgca) [inline electron/api/index.ts:25]: This procedure takes a property called `name`, and returns "Hello {name}"

File: electron/api/index.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        emit.next({ text })
+      }
+
+      ee.on('demo-event', onGreet)
+
+      return () => {
+        ee.off('demo-event', onGreet)
+      }
+    })
+  }),
```
Comment (dgca) [inline electron/api/index.ts:38]: This is a subscription procedure, and it'll send `{ text: "Count is {count}" }` to any subscribers.

File: electron/api/index.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    if (!ifNode) {
+      throw new Error('Node not foound')
+    }
+
+    const accounts = await ifNode.wallet.listAccounts()
+
+    return accounts.map(account => {
+      return account.publicAddress
+    })
+  }),
```
Comment (dgca) [inline electron/api/index.ts:49]: This is to show that we can actually return node data.

File: electron/common/index.ts [hunk 6]
```diff
@@ -1,5 +1,3 @@
-/* eslint-disable @typescript-eslint/ban-ts-comment */
-//@ts-nocheck
```
Comment (dgca) [inline electron/common/index.ts:2]: lol, why tho

File: electron/ironfish/IronFishManager.ts [hunk 7]
```diff
@@ -35,6 +35,8 @@ import sendMessageToRender from '../utils/sendMessageToRender'
 import { WorkerMessageType } from '@ironfish/sdk/build/src/workerPool/tasks/workerMessage'
 import { app, shell } from 'electron'
 
+export let ifNode: null | FullNode = null
```
Comment (dgca) [inline electron/ironfish/IronFishManager.ts:38]: This is a messy way to expose the node to the tRPC server, but it works 🤷  Could clean this up if we decide to go with this.

File: src/App.tsx [hunk 8]
```diff
@@ -45,46 +46,51 @@ const queryClient = new QueryClient()
...
-                  <Route
-                    path={ROUTES.ADDRESS_BOOK_DETAILS}
-                    element={<AddressDetails />}
-                  />
-                  {/* <Route path={ROUTES.RESOURCES} element={null} /> */}
-                  <Route path={ROUTES.NODE} element={<NodeOverview />} />
-                  <Route element={<ReleaseNotesProviderWrapper />}>
-                    <Route path={ROUTES.UPDATE} element={<Update />} />
-                    <Route path={ROUTES.UPDATES} element={<Updates />} />
+      <TRPCProvider>
```
Comment (dgca) [inline src/App.tsx:49]: Just adding a provider to the React app.

File: src/components/TRPCDemo/TRPCDemo.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
+import { trpcReact } from 'Providers/TRPCProvider'
```
Comment (dgca) [inline src/components/TRPCDemo/TRPCDemo.tsx:1]: And this file demos how the client actually gets data from the server.

File: src/components/TRPCDemo/TRPCDemo.tsx [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  if (isError) {
+    return <h1>Error: {error?.message}</h1>
+  }
+
+  if (!data) {
+    return null
+  }
+
+  return <h1>{data.text}</h1>
+}
```
Comment (dgca) [inline src/components/TRPCDemo/TRPCDemo.tsx:32]: Here I wanted to show how the tRPC client uses react-query, so all procedure calls come with built in loading/error/pending/etc states, which is super nice. We wouldn't have to use stuff like the custom `useAsyncDataWrapper` anymore.

File: src/components/TRPCDemo/TRPCDemo.tsx [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+function Subscription() {
+  const [subscriptionData, setSubscriptionData] = useState<string | null>(null)
+  trpcReact.subscriptionDemo.useSubscription(undefined, {
+    onData: data => {
+      setSubscriptionData(data?.text || null)
+    },
+  })
+
+  return <h1>Data: {subscriptionData}</h1>
+}
```
Comment (dgca) [inline src/components/TRPCDemo/TRPCDemo.tsx:43]: Here's that subscription demo. This is updating state every 5 seconds with the new message from the backend. AFAIK, we don't have subscriptions in the app currently, and just rely on polling? Not sure about that, but either way, this is how tRPC does subscriptions.

File: src/components/TRPCDemo/TRPCDemo.tsx [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+function Accounts() {
+  const { data } = trpcReact.getAccounts.useQuery()
+
+  if (!data) {
+    return null
+  }
+
+  return <div>{JSON.stringify(data, null, 2)}</div>
+}
```
Comment (dgca) [inline src/components/TRPCDemo/TRPCDemo.tsx:53]: And here we're demoing getting node data from the server.

File: src/providers/TRPCProvider.tsx [hunk 13]
```diff
@@ -0,0 +1,23 @@
+import { useState, ReactNode } from 'react'
```
Comment (dgca) [inline src/providers/TRPCProvider.tsx:1]: This is the provider I mentioned earlier. It just sets up the system and exports the typed `trpcReact` utility that we're using to actually make the procedure calls.

File: electron/api/index.ts [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+})
+
+export type AppRouter = typeof router
```
Comment (dgca) [inline electron/api/index.ts:52]: This is the magic that makes the client code type-safe 🪄

**Comment by dguenther** [comment]
Closing as this version is no longer in development

**Pull request closed without merging**